packages/walletconnect/src/index.ts

```diff
@@ -73,47 +67,57 @@ export class WalletConnect extends Connector {
     if (error) this.onError?.(error)
   }
 
-  private chainChangedListener = (chainId: number | string): void => {
-    this.actions.update({ chainId: parseChainId(chainId) })
+  private chainChangedListener = (chainId: number): void => {
+    this.actions.update({ chainId })
   }
 
   private accountsChangedListener = (accounts: string[]): void => {
     this.actions.update({ accounts })
   }
 
-  private URIListener = (_: Error | null, payload: { params: string[] }): void => {
-    this.events.emit(URI_AVAILABLE, payload.params[0])
+  private URIListener = (uri: string): void => {
+    this.events.emit(URI_AVAILABLE, uri)
   }
 
-  private async isomorphicInitialize(chainId = this.defaultChainId): Promise<void> {
-    if (this.eagerConnection) return
-
-    // because we can only use 1 url per chainId, we need to decide between multiple, where necessary
-    const rpc = Promise.all(
-      Object.keys(this.rpc).map(
-        async (chainId): Promise<[number, string]> => [
-          Number(chainId),
-          await getBestUrl(this.rpc[Number(chainId)], this.timeout),
-        ]
-      )
-    ).then((results) =>
-      results.reduce<{ [chainId: number]: string }>((accumulator, [chainId, url]) => {
-        accumulator[chainId] = url
-        return accumulator
-      }, {})
-    )
-
-    return (this.eagerConnection = import('@walletconnect/ethereum-provider').then(async (m) => {
-      this.provider = new m.default({
+  private async isomorphicInitialize(desiredChainId: number | undefined = this.defaultChainId): Promise<MockWalletConnectProvider> {
+    if (this.eagerConnection) return this.eagerConnection
+
+    if (this.provider) {
+      await this.deactivate()
+    }
+
+    const rpcMap = this.rpcMap
+    const rpcBestUrlMap = rpcMap
+      ? Object.fromEntries(
+          await Promise.all(
+            Object.entries(rpcMap).map(
+              async ([chainId, map]): Promise<[string, string]> => [`${chainId}`, await getBestUrl(map, this.timeout)]
+            )
+          )
+        )
+      : undefined
+
+    const chains = [...this.chains]
+    if (desiredChainId) {
+      const idx = chains.indexOf(desiredChainId)
+      if (idx === -1) {
+        throw new Error(`Invalid chainId ${desiredChainId}. Make sure to include it in the "chains" option.`)
+      }
+      chains.splice(idx, 1)
+      chains.unshift(desiredChainId)
+    }
+
+    return (this.eagerConnection = import('@walletconnect/ethereum-provider').then(async (ethProviderModule) => {
+      const provider = (this.provider = (await ethProviderModule.default.init({
```

Comment:
Will try! That would be indeed quite helpful!

Reply:
Unfortunately, `on` is of type `void`, so no chaining here :< And the assignment to locally-scoped variable is to avoid safe property access - in theory, I don't think there's any chance this can get nullified - we're not in a multithreaded environment.